bundles/org.eclipse.core.resources/src/org/eclipse/core/internal/events/BuildManager.java

```diff
@@ -1272,22 +1273,48 @@ private boolean needsBuild(InternalBuilder builder, int trigger) {
 		});
 
 		//search for the builder's project
-		if (currentDelta.findNodeAt(builder.getProject().getFullPath()) != null) {
-			if (Policy.DEBUG_BUILD_NEEDED)
-				Policy.debug(toString(builder) + " needs building because of changes in: " + builder.getProject().getName()); //$NON-NLS-1$
-			return true;
+		boolean ret = false;
+		AbstractDataTreeNode found;
+		if ((found = currentDelta.findNodeAt(builder.getProject().getFullPath())) != null) {
+			if (Policy.DEBUG_BUILD_NEEDED) {
+				String prefix = toString(builder);
+				Policy.debug(prefix + " needs building because of changes in: " + builder.getProject().getName()); //$NON-NLS-1$
+				if (debugBuildNeededDelta())
+					debugDeltaNode(prefix, found);
+			}
+			ret = true;
+			if (!debugBuildNeededDelta())
+				return ret;
 		}
 
 		//search for builder's interesting projects
 		IProject[] projects = builder.getInterestingProjects();
 		for (IProject project : projects) {
-			if (currentDelta.findNodeAt(project.getFullPath()) != null) {
-				if (Policy.DEBUG_BUILD_NEEDED)
-					Policy.debug(toString(builder) + " needs building because of changes in: " + project.getName()); //$NON-NLS-1$
-				return true;
+			if ((found = currentDelta.findNodeAt(project.getFullPath())) != null) {
+				if (Policy.DEBUG_BUILD_NEEDED) {
+					String prefix = toString(builder);
+					Policy.debug(prefix + " needs building because of changes in: " + project.getName()); //$NON-NLS-1$
+					if (debugBuildNeededDelta())
+						debugDeltaNode(prefix, found);
+				}
+				ret = true;
+				if (!debugBuildNeededDelta())
+					return ret;
 			}
 		}
-		return false;
+		return ret;
+	}
+
+	private boolean debugBuildNeededDelta() {
+		return Policy.DEBUG_BUILD_NEEDED && Policy.DEBUG_BUILD_NEEDED_DELTA;
```

Comment:
why should both flags needed to be set?

Reply:
The header line "needs building because of changes in" is (and should remain) controlled by DEBUG_BUILD_NEEDED and does make sense without the details, while the delta details make little sense without that header line.
So combining the flags feels necessary in one ("or", for the header) or another ("and", for the details) way.